store/inputValidation.go

```diff
@@ -0,0 +1,86 @@
+package store
+
+import (
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/xmidt-org/argus/model"
+)
+
+var (
+	idFormatRegex     *regexp.Regexp
+	bucketFormatRegex *regexp.Regexp
+)
+
+var (
+	errInvalidID               = BadRequestErr{Message: "Invalid ID format. Expecting the format of a SHA-256 message digest."}
+	errIDMismatch              = BadRequestErr{Message: "IDs must match between the URL and payload."}
+	errDataFieldMissing        = BadRequestErr{Message: "Data field must be set in payload."}
+	errBodyReadFailure         = BadRequestErr{Message: "Failed to read body."}
+	errPayloadUnmarshalFailure = BadRequestErr{Message: "Failed to unmarshal json payload."}
+	errInvalidBucket           = BadRequestErr{Message: "Invalid bucket format."}
+)
+
+func init() {
+	idFormatRegex = regexp.MustCompile(`^[0-9a-f]{64}$`)
+	bucketFormatRegex = regexp.MustCompile("^[0-9a-z][0-9a-z-]{1,61}[0-9a-z]$")
+}
+
+func validateItemTTL(item *model.Item, maxTTL time.Duration) {
+	ttlMaxSeconds := int64(maxTTL.Seconds())
+	if item.TTL == nil || *item.TTL > ttlMaxSeconds {
+		item.TTL = &ttlMaxSeconds
+	}
+}
+
+// normalizeID should be run on all instances of item IDs decoded from incoming requests.
+func normalizeID(ID string) string {
+	return strings.ToLower(strings.TrimSpace(ID))
+}
```

Comment:
Instead of having to remember to do this, you could have the ID be a special `type` with its own `UnmarshalText()` function, where you can run this function as part of unmarshalling; then all unmarshalling automatically include this logic.  Could also just make a special UnmarshalJSON() function for the whole store.Item to handle TTL and others...

I believe the ID is sometimes in the URL? My proposals above wouldn't solve for that.

Reply:
`UnmarshalText()` would/could also handle `isIDValid()`.

Reply:
Yeah, ID comes both from the URL path and the body and we check they're equal once we're normalized them. Let me take a look at that. Maybe, we can create an ItemJSONUnmarshaler of some sorts that could at least ensure the ID is normalized.